### PR TITLE
Add support for avatar prop to the Alert component

### PR DESCRIPTION
**Pull request by rohitpagariya**
This PR adds support to render an Avatar component in place of an icon in the Alert component. 

And thank you @ling1726 for adding the bundle size fixture for the Alert component :)

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/fd4c88b3915ed3e26d4d39bcb24390343b8299ac'>fd4c88b3915ed3e26d4d39bcb24390343b8299ac</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/Avatar-mount_base.html">1840</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/Avatar-mount.html">1825</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/Button-mount_base.html">1249</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/Button-mount.html">1306</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProvider-mount_base.html">2787</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProvider-mount.html">2667</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProviderWithTheme-mount_base.html">930</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProviderWithTheme-mount.html">915</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">893</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">928</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">928</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">1007</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/MakeStyles-mount_base.html">2447</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/23591/perf-test-react-components/MakeStyles-mount.html">2460</a></td>
            <td>50000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Review by ling1726 — DISMISSED** [review]
The anatomy of the `Toast` (which I'm going by because there is no `Alert` figma spec) has no slot for an avatar. Likewise there is no mention of `avatar` slot in the component spec.

Do you mind clarifying what the requirements are for this slot ? Since the design requirements use icons and avatar interchangeably does it make sense to rename the existing icon slot to something like `media` (I see there are no avatar defaults in this PR)? or are there some specific style overrides for both cases ?

**Comment by ling1726** [comment]
> @ling1726 you find the Toast figma here - https://www.figma.com/file/HQ4nlkneIZvWUFzehrQoEL/Toast?node-id=1866%3A221039
> 
> It calls out intent="info-avatar" as the prop to render an avatar. Of course, the consumer also needs to pass in the avatar prop. So I decided to simplify the api by assuming that if avatar prop is present, Avatar needs to be rendered in place of an icon.
> 
> Unfortunately, the design spec is not clear on the interplay between icon/intent/avatar props. I think we, the dev team, has some leeway to design the api surface as needed.

> I think we, the dev team, has some leeway to design the api surface as needed. 

Are there any meaningful default props that need to be added to the avatar or icon components to follow the design spec ? 

The biggest downside to taking a direct dependency on `Avatar` is bundle size. The component itself has a very large bundle size due to the amount of styles it has. Therefore anyone using the `Alert` component will get the avatar in their bundle

File: packages/react-components/react-alert/src/components/Alert/Alert.test.tsx [hunk 1]
```diff
@@ -41,6 +41,12 @@ describe('Alert', () => {
     expect(screen.getByTestId('foo')).toBeTruthy();
   });
 
+  it('renders an avatar', () => {
+    render(<Alert avatar={{ name: 'John Doe' }}>Test</Alert>);
+    screen.debug();
```
Comment (layershifter) [inline packages/react-components/react-alert/src/components/Alert/Alert.test.tsx:46]:
```suggestion
```

File: packages/react-components/react-alert/Spec.md [hunk 2]
```diff
@@ -38,6 +38,7 @@ See [Alert.types.ts](./src/components/Alert/Alert.types.ts)
 - `root`: The outermost `<div>` that contains the rest of the slots
 - `icon`: (optional) A `<span>` that renders an icon and is inferred by the `intent` prop or passed in via the `icon` prop
 - `action`: (optional) A `<button>` that prompts users to act on it
+- `avatar`: (optional) A `<span>` that renders an Avatar component passed in via the `avatar` prop
```
Comment (layershifter) [inline packages/react-components/react-alert/Spec.md:41]:
```suggestion
- `avatar`: (optional) A slot that renders an `Avatar` component passed in via the `avatar` prop
```

It's not the span, it's an actual `Avatar`

File: packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import * as React from 'react';
+
+import { DismissCircleRegular } from '@fluentui/react-icons';
```
Comment (layershifter) [inline packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx:3]:
```suggestion
import * as React from 'react';
import { DismissCircleRegular } from '@fluentui/react-icons';
```

File: packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+import { Alert } from '../../index';
+
+export const AvatarProp = () => (
```
Comment (layershifter) [inline packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx:7]:
```suggestion
export const Avatar = () => (
```

nit: The name should match the filename without the component name prefix

File: packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        appearance: 'transparent',
+        children: 'Undo',
+      }}
+      avatar={{ name: 'Jane Doe', size: 24, color: 'dark-red' }}
+    >
+      sent you a message
+    </Alert>
+    <Alert
+      action={{
+        appearance: 'transparent',
```
Comment (layershifter) [inline packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx:20]:
```suggestion
```

I overlooked this in the previous PR, but let's move `appearance` to `defaultProps` of the slot so users will not need to specify it by default:

```diff
-action: resolveShorthand(props.action),
+action: resolveShorthand(props.action, {
+  defaultProps: { appearance: 'transparent' },
+}),
```

Please also update other `Alert` stories.

File: packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        icon: <DismissCircleRegular />,
+      }}
+      avatar={{ name: 'John Smith', size: 24, color: 'dark-green' }}
```
[the remark this answers is not in the document]
Comment (rohitpagariya) [reply packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx:23]: The default size is 32px hence the additional param

File: packages/react-components/react-alert/src/stories/Alert/index.stories.tsx [hunk 7]
```diff
@@ -2,8 +2,9 @@ import { Alert } from '@fluentui/react-alert';
 import descriptionMd from './AlertDescription.md';
 
 export { Default } from './AlertDefault.stories';
-export { Icon } from './AlertIcon.stories';
 export { Intent } from './AlertIntent.stories';
+export { Icon } from './AlertIcon.stories';
+export { AvatarProp } from './AlertAvatar.stories';
```
[the remark this answers is not in the document]
Comment (rohitpagariya) [reply packages/react-components/react-alert/src/stories/Alert/index.stories.tsx:7]: I remember this causing type errors, but will try again.

File: packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx (hunk 5 above)
Comment (rohitpagariya) [reply packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx:20]: Great suggestion, I didn't know this was possible. Thank you!

File: packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx (hunk 4 above)
Comment (rohitpagariya) [reply packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx:7]: I changed it due to a type error as I was importing the actual Avatar component. But that's not the case anymore so fixed it :)

File: packages/react-components/react-alert/src/components/Alert/Alert.test.tsx (hunk 1 above)
Comment (rohitpagariya) [reply packages/react-components/react-alert/src/components/Alert/Alert.test.tsx:46]: Good catch, removed

File: change/@fluentui-react-alert-640db5c1-33ff-4314-b504-1fb3b169e05f.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "prerelease",
+  "comment": "Add avatar support to Alert",
```
Comment (khmakoto) [inline change/@fluentui-react-alert-640db5c1-33ff-4314-b504-1fb3b169e05f.json:3]:
```suggestion
  "comment": "feat: Adding avatar slot support to Alert.",
```

File: packages/react-components/react-alert/src/components/Alert/Alert.types.ts [hunk 9]
```diff
@@ -15,6 +16,10 @@ export type AlertSlots = {
    * The action slot renders a button that prompts the user to take action on the alert
    */
   action?: Slot<typeof Button>;
+  /**
+   * The avatar slot renders the Avatar component passed in as a prop
```
Comment (khmakoto) [inline packages/react-components/react-alert/src/components/Alert/Alert.types.ts:20]: I think a better description would be something like

```suggestion
   * The avatar slot renders an avatar before the contents of the alert.
```

**Review by khmakoto — COMMENTED** [review]
Can you note in your PR description that you are also adding the bundle size fixture for the Aler component in this PR?

**Review by khmakoto — APPROVED** [review]

File: packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
+import { DismissCircleRegular } from '@fluentui/react-icons';
+
+import { Alert } from '../../index';
```
Comment (layershifter) [inline packages/react-components/react-alert/src/stories/Alert/AlertAvatar.stories.tsx:4]:
```suggestion
import { Alert } from '@fluentui/react-alert';
```

**Review by layershifter — APPROVED** [review]

**Pull request merged**